pkgs/tools/networking/pritunl-client/default.nix

```
@@ -1,27 +1,144 @@
-{ lib, buildGoModule, fetchFromGitHub }:
-
-buildGoModule rec {
-  pname = "pritunl-client";
+{ lib
+, stdenv
+, fetchFromGitHub
+, runtimeShell
+, runCommand
+, makeWrapper
+, installShellFiles
+, buildGoModule
+, coreutils
+, which
+, gnugrep
+, gnused
+, openresolv
+, systemd
+, iproute2
+, openvpn
+, electron
+, gcc-unwrapped
+}: let
   version = "1.3.3484.2";
-
   src = fetchFromGitHub {
     owner = "pritunl";
     repo = "pritunl-client-electron";
     rev = version;
     sha256 = "sha256-thIbw/Iek2vNNmkOBHrzEhTbaOm05CsbjqORQKC2jIs=";
   };
 
-  modRoot = "cli";
-  vendorHash = "sha256-miwGLWpoaavg/xcw/0pNBYCdovBnvjP5kdaaGPcRuWk=";
+  cli = buildGoModule {
+    pname = "pritunl-cli";
+    inherit version src;
+
+    modRoot = "cli";
+    vendorHash = "sha256-miwGLWpoaavg/xcw/0pNBYCdovBnvjP5kdaaGPcRuWk=";
+
+    postInstall = ''
+      mv $out/bin/cli $out/bin/pritunl-client
+    '';
+  };
+
+  service = buildGoModule rec {
+    pname = "pritunl-client-service";
+    inherit version src;
+
+    modRoot = "service";
+    vendorHash = "sha256-VuJ1giVSDlVFRJOy0UPVYp8j6TCWzXbiTgg4xyMq3dQ=";
+
+    nativeBuildInputs = [makeWrapper];
+
+    buildInputs = [
+      coreutils
+      which
+      gnused
+      gnugrep
+    ] ++ lib.optionals stdenv.isLinux [
+      openresolv
+      systemd
+      iproute2
+    ];
+
+    postPatch = ''
+      sed -i -E 's|#!\s*(/usr)?/bin/(env )?bash\b|#! ${runtimeShell}|g' service/profile/scripts.go
+    ''
+    + lib.optionalString stdenv.isLinux ''
+      sed -i -E 's|(/usr)?/s?bin/busctl\b|busctl|g' service/profile/scripts.go
+      sed -i -E 's|(/usr)?/s?bin/resolvectl\b|resolvectl|g' service/profile/scripts.go
+      sed -i -E 's|(/usr)?/s?bin/ip\b|ip|g' service/profile/scripts.go
+    '';
+
+    postInstall = ''
+      mv $out/bin/service $out/bin/pritunl-client-service
+    '' + lib.optionalString stdenv.isLinux ''
+      mkdir -p $out/lib/systemd/system/
+      cp $src/resources_linux/pritunl-client.service $out/lib/systemd/system/
+    '';
+
+    postFixup = let
+      openvpn-wrapped = runCommand "openvpn-wrapped" {nativeBuildInputs = [makeWrapper];} ''
+        mkdir -p $out/bin
+        makeWrapper ${openvpn}/bin/openvpn $out/bin/openvpn \
+          --prefix PATH : ${lib.makeBinPath buildInputs} \
+          --add-flags "--setenv PATH \$PATH"
```

```suggestion
      openvpn-wrapped = runCommand "openvpn-wrapped" {
        nativeBuildInputs = [ makeWrapper ];
      } ''
        mkdir -p $out/bin
        makeWrapper ${openvpn}/bin/openvpn $out/bin/openvpn \
          --prefix PATH : ${lib.makeBinPath buildInputs} \
          --add-flags "--setenv PATH \$PATH"
```
Why do we need to wrap openvpn? Can't we add this to the systemd service file's path? Or does this need to be fixed within openvpn itself?

---

`pritunl-client-service` picks `openvpn` from its PATH, but openvpn doesn't pass down its PATH to the hook scripts (where the sh utils are needed), so I need to wrap to add the `--setenv PATH $PATH` argument, and at the same time also include the shell utils in its path to ensure they're not overriden. Also, the scripts are hardcoded.

Technically, I would not need those utils in `pritunl-client-service` or systemd unit directly, only on openvpn, but decided to keep it in `buildInputs` for clarity.

---

keeping those things in buildInputs is wrong and not how it is supposed to be used. Also please rather patch openvpn directly as this is not an issue that is purely on pritunl.

---

Ok, I can take those out of `buildInputs`.

But I'd say it's not right to patch `openvpn`; not passing environment variables to hook scripts is the expected/well documented/upstream behavior of openvpn, and changing it could break other stuff which depend on this behavior.

Just Pritunl specifically expect the shell utils it hardcodes in the hook scripts to be available on system shell's default PATH, but of course this doesn't work for NixOS, so we need openvpn to be wrapped only for this case, to make the store tools we choose be passed to hook scripts.

An alternative would be to substitute the shebangs from `runtimeShell` to some `bash` derivation/wrapper which exports PATH before running the script. Then default openvpn would work. Is there some common idiom for this interpreter? Or would this be achieved with a normal wrapper for bash, and using it as shebang interpreter for the hook scripts?

---

@SuperSandro2000 Moved script dependencies to a variable, out of `buildInputs`.

---

> But I'd say it's not right to patch `openvpn`; not passing environment variables to hook scripts is the expected/well documented/upstream behavior of openvpn, and changing it could break other stuff which depend on this behavior.

But not passing down PATH is in the context of nixos rather severe because you almost everything.

> Just Pritunl specifically expect the shell utils it hardcodes in the hook scripts to be available on system shell's default PATH, but of course this doesn't work for NixOS, so we need openvpn to be wrapped only for this case, to make the store tools we choose be passed to hook scripts.

we can't patch that and just amend our paths to that? Or why don't we patch the scrips themselves?

> An alternative would be to substitute the shebangs from `runtimeShell` to some `bash` derivation/wrapper which exports PATH before running the script. Then default openvpn would work. Is there some common idiom for this interpreter? Or would this be achieved with a normal wrapper for bash, and using it as shebang interpreter for the hook scripts?

We could use the binaryWrapper for that but we should first see if we can't avoid that.

---

The scripts are not in store, they're hardcoded and written dynamically to `/tmp` by the go code; I've patched the shebang there, but trying to figure out every utils it calls is really error-prone, if/when they update those scripts in some update. I think this providing some basic utils in PATH is the best we can do for now, as it's done only in the wrapped executable anyway.